modules/substrate/src/lib.rs

```diff
@@ -375,8 +380,18 @@ impl<T: Config> bp_header_chain::HeaderChain<BridgedHeader<T>, sp_runtime::Dispa
 		PalletStorage::<T>::new().current_authority_set()
 	}
 
-	fn append_header(header: BridgedHeader<T>) {
+	fn append_header(header: BridgedHeader<T>) -> Result<(), sp_runtime::DispatchError> {
+		// We do a quick check here to ensure that our header chain is making progress and isn't
+		// "travelling back in time" (which would be indicative of something bad, e.g a hard-fork).
+		//
+		// Note: It is valid for GRANDPA to finalize the same header twice as long as it is
+		// finalized in different rounds. This is why we don't check for a strictly greater-than
+		// relationship here.
+		let best_finalized = PalletStorage::<T>::new().best_finalized_header().header;
+		ensure!(best_finalized.number() <= header.number(), <Error<T>>::ConflictingFork);
```

Comment:
Imo we shouldn't care about what happens in GRANDPA. The pallet should care whether transaction is improving its state. If transaction brings the same header and (the same or other) justification again, why should we go further - we'll end up with the same state (but with block that has been executed slower than it may be)?

Reply:
Also two points:
1) malicious relayer may just submit the same header+justification over and over again to increase `RequestCount` counter value, thus blocking valid transactions to be mined and new headers to be finalized;
2) iiuc the second import of the header that schedules+enacts authorities set change will end up at [this line](https://github.com/paritytech/parity-bridges-common/blob/master/modules/substrate/src/lib.rs#L409). So we'll increase the `set_id` twice => the pallet would stuck.

Reply:
I second Slava on this. Importing the same header twice really doesn't bring any new information. Let's be strict here.

Reply:
> Imo we shouldn't care about what happens in GRANDPA. The pallet should care whether transaction is improving its state. If transaction brings the same header and (the same or other) justification again, why should we go further - we'll end up with the same state (but with block that has been executed slower than it may be)?

So _technically_ it's not the same state since GRANDPA has made progress by finalizing the same block in different rounds. However, from the perspective of the bridge you're right, we shouldn't care about this detail since we want "real" progress.

> iiuc the second import of the header that schedules+enacts authorities set change will end up at this line. So we'll increase the set_id twice => the pallet would stuck.

That's a good point.  In the future we'd ideally be getting the `set_id` directly from the digest, so this could be avoided. However, in its current form this doesn't work.